src/t_set.c

```diff
@@ -1093,7 +1097,7 @@ void sunionDiffGenericCommand(client *c, robj **setkeys, int setnum,
         algo_one_work /= 2;
         diff_algo = (algo_one_work <= algo_two_work) ? 1 : 2;
 
-        if (diff_algo == 1 && setnum > 1) {
+        if (diff_algo == 1 && setnum > 1 && !sameset) {
```

Comment:
isn't that change unnecessary (we won't reach here since the outer `if` will be false)